### ISSUE TITLE
Play 2.6.5 does not work with Java 9

### ISSUE THREAD

**Issue by markehammons** [playframework/playframework#7879]
### Play Version (2.5.x / etc)

 2.6.5

### API (Scala / Java / Neither / Both)

Scala

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10) 
Fedora 26

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

openjdk version "9.0.0.15"
OpenJDK Runtime Environment (Zulu build 9.0.0.15+181)
OpenJDK 64-Bit Server VM (Zulu build 9.0.0.15+181, mixed mode)

### Expected Behavior

I expect my application to load!

### Actual Behavior

```
java.lang.NumberFormatException: Not a version: 9
        at scala.util.PropertiesTrait$class.parts$1(Properties.scala:176)
        at scala.util.PropertiesTrait$class.isJavaAtLeast(Properties.scala:180)
        at scala.util.Properties$.isJavaAtLeast(Properties.scala:16)
        at play.dev.filewatch.FileWatchService$$anon$1.delegate$lzycompute(FileWatchService.scala:68)
        at play.dev.filewatch.FileWatchService$$anon$1.delegate(FileWatchService.scala:66)
        at play.dev.filewatch.FileWatchService$$anon$1.watch(FileWatchService.scala:79)
        at play.runsupport.Reloader.<init>(Reloader.scala:359)
        at play.runsupport.Reloader$.reloader$lzycompute$1(Reloader.scala:220)
        at play.runsupport.Reloader$.play$runsupport$Reloader$$reloader$1(Reloader.scala:220)
        at play.runsupport.Reloader$.startDevMode(Reloader.scala:230)
        at play.sbt.run.PlayRun$$anonfun$playRunTask$1$$anonfun$apply$2$$anonfun$apply$3.devModeServer$lzycompute$1(PlayRun.scala:77)
        at play.sbt.run.PlayRun$$anonfun$playRunTask$1$$anonfun$apply$2$$anonfun$apply$3.play$sbt$run$PlayRun$$anonfun$$anonfun$$anonfun$$devModeServer$1(PlayRun.scala:77)
        at play.sbt.run.PlayRun$$anonfun$playRunTask$1$$anonfun$apply$2$$anonfun$apply$3.apply(PlayRun.scala:100)
        at play.sbt.run.PlayRun$$anonfun$playRunTask$1$$anonfun$apply$2$$anonfun$apply$3.apply(PlayRun.scala:63)
        at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
```

### Reproducible Test Case

I imagine this is reproducible with any application if you're using a java 9 vm. It appears there is a hardcoded list of java versions in the framework, and 9 is not in the list.

**Comment by jroper** [comment]
I imagine this will be the easiest issue to fix of all the JDK9 compatibility issues that we're going to encounter.

**Comment by richdougherty** [comment]
Thanks for the report @markehammons. Agree it would be good to get working with Java 9!

**Comment by marcospereira** [comment]
Hey @markehammons, which version of Scala is your app is using?

I was able to create a play app as I described here: https://groups.google.com/forum/#!topic/play-framework/e3KLg94j5_M

Could you please try with Scala 2.12.3 and report back if it works for you?

**Comment by markehammons** [comment]
I'm using scala 2.12.3

**Comment by marcospereira** [comment]
Hi @markehammons,

The Scala issue was https://github.com/scala/scala/pull/5276 and it was released in 2.12.1 and [preliminary suppor](https://github.com/sbt/sbt/blob/v0.13.14-RC1/notes/0.13.14.markdown#improvements) was added in sbt 0.13.14.

Is it possible to create a small project to reproduces the problem? I'm running the following build of Java 9:

```
java version "9"
Java(TM) SE Runtime Environment (build 9+181)
Java HotSpot(TM) 64-Bit Server VM (build 9+181, mixed mode)
```

**Comment by Spenhouet** [comment]
I had the same issue. It's just not possible to use Play with Java 9 at the moment.
Scala doesn't support the released Java 9 as of yet:
https://github.com/scala/scala-dev/issues/139

It seems that they are far from ready and I don't see any progress.

**Comment by marcospereira** [comment]
> I had the same issue. While this one issue got fixed there are many others. It's just not possible to use Play with Java 9 at the moment.

Hi @Spenhouet, what other issues did you find?

Java support will probably come in two waves here, one to ensure that everything runs fine on Java 9 and other to use new features (like the module system) and APIs. We first need to ensure that Play applications can run fine on Java 9. The same is happening with sbt and Scala, that is why sbt support in 0.13.14 was described as preliminary and see how sbt/sbt#3587 is also described as "improved" compatibility. 

So, what issues did you find? What is making not possible to use Play with Java 9 at the moment? Is it possible to reproduce them?

Thanks.

**Comment by Spenhouet** [comment]
@marcospereira Oh, it seems I had a different issue. After looking up and solving the issue described in this thread I got a new error.
`scala.reflect.internal.MissingRequirementError: object java.lang.Object in compiler mirror not found.`
At this point I gave up and blamed it on the long list of tasks for the Java 9 support for scala (https://github.com/scala/scala-dev/issues/139).

I just found out that somehow my IDE (IntelliJ) does ignore the `scalaVersion` defined in `build.sbt` and just loads the default scala version for sbt version 0.13.16 that is scala 2.10.6.
I don't know if this is a error of IntelliJ or the Play plugin.
It looks like I'm not the only one with this issue: https://stackoverflow.com/questions/45305922/why-intellij-idea-uses-scala-2-10-6-for-play2-configuration

Building my Play project via the sbt command line with Java 9, scala 2.12.3 and sbt 0.13.16 does work fine.

This is probably solved when sbt uses a new default scala version (https://github.com/sbt/sbt/issues/3587) and play released the support for sbt 1.x.x (https://github.com/playframework/playframework/issues/7261).

**Comment by marcospereira** [comment]
Thank you for the feedback, @Spenhouet.

You can try with one of the nightly built snapshot, for example, this version: `2.6.6-2017-10-05-829d140-SNAPSHOT`. You will need the following resolver too:

```scala
resolvers += Resolver.sonatypeRepo("snapshots")
```
And of course change `build.properties` to use sbt 1.0.2.

Can you check and see if it works?

**Comment by Spenhouet** [comment]
@marcospereira I've tried that and is resolves the error

`[error] unresolved dependency: com.typesafe.play#sbt-plugin;2.6.5: not found`

but it still leaves me with the following errors:

```
[error] sbt.librarymanagement.ResolveException: unresolved dependency: com.typesafe.sbt#sbt-jshint;1.0.5: not found
[error] unresolved dependency: com.typesafe.sbt#sbt-rjs;1.0.9: not found
[error] unresolved dependency: com.typesafe.sbt#sbt-play-enhancer;1.2.1: not found
```
Are there snapshot versions of these libraries two?

**Comment by marcospereira** [comment]
@Spenhouet, you are right. There are still work to be done for sbt-jshint, sbt-rjs and play-enhancer. We really appreciate if you can help here.

**Comment by markehammons** [comment]
after upgrading to 2.6.6 and sbt 1.0.2, play now launches my app, but it crashes while rendering the home page with this error: 

java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter
        at io.jsonwebtoken.impl.Base64Codec.decode(Base64Codec.java:26)
        at io.jsonwebtoken.impl.DefaultJwtBuilder.signWith(DefaultJwtBuilder.java:106)
        at play.api.mvc.JWTCookieDataCodec$JWTFormatter.format(Cookie.scala:718)
        at play.api.mvc.JWTCookieDataCodec.encode(Cookie.scala:577)
        at play.api.mvc.JWTCookieDataCodec.encode$(Cookie.scala:575)
        at play.api.mvc.DefaultJWTCookieDataCodec.encode(Cookie.scala:768)
        at play.api.mvc.FallbackCookieDataCodec.encode(Cookie.scala:742)
        at play.api.mvc.FallbackCookieDataCodec.encode$(Cookie.scala:741)
        at play.api.mvc.DefaultSessionCookieBaker.encode(Session.scala:95)
        at play.api.mvc.CookieBaker.encodeAsCookie(Cookie.scala:414)
        at play.api.mvc.CookieBaker.encodeAsCookie$(Cookie.scala:413)
        at play.api.mvc.DefaultSessionCookieBaker.encodeAsCookie(Session.scala:95)
        at play.api.mvc.Result.$anonfun$bakeCookies$2(Results.scala:297)
        at scala.Option.map(Option.scala:146)
        at play.api.mvc.Result.bakeCookies(Results.scala:296)
        at play.core.server.common.ServerResultUtils.prepareCookies(ServerResultUtils.scala:227)
        at play.core.server.AkkaHttpServer.$anonfun$executeAction$3(AkkaHttpServer.scala:302)
        at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
        at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:55)
        at akka.http.scaladsl.util.FastFuture$.flatMap$extension(FastFuture.scala:26)
        at play.core.server.AkkaHttpServer.executeAction(AkkaHttpServer.scala:301)
        at play.core.server.AkkaHttpServer.executeHandler(AkkaHttpServer.scala:254)
        at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:200)
        at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$3(AkkaHttpServer.scala:106)
        at akka.stream.impl.fusing.MapAsync$$anon$23.onPush(Ops.scala:1172)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:499)
        at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:462)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:368)
        at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:571)
        at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:457)
        at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:546)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:725)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:740)
        at akka.actor.Actor.aroundReceive(Actor.scala:514)
        at akka.actor.Actor.aroundReceive$(Actor.scala:512)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:650)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)
        at akka.actor.ActorCell.invoke(ActorCell.scala:496)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

**Comment by markehammons** [comment]
Using the solution here:

> Hey @markehammons, which version of Scala is your app is using? I was able to create a play app as I described here: https://groups.google.com/forum/#!topic/play-framework/e3KLg94j5_M Could you please try with Scala 2.12.3 and report back if it works for you?

I was able to get my application to load and run (at least for all basic tasks). So the next question is can 
the module adding be done without changing sbt? I tried adding javaOptions += "--add-modules=..." to my build and fork := true, and unfortunately play still crashes unless i export the option before i run sbt.

**Comment by Neverminder** [comment]
I'm facing the same issue. With `javacOptions += "--add-modules=java.xml.bind"` added I can now `sbt compile`, however `javaOptions in Universal += "--add-modules=java.xml.bind"` doesn't do anything (tried in dev mode), even though application launches, it crashes with the first http request with `Uncaught error from thread [play-dev-mode-akka.actor.default-dispatcher-3]: javax/xml/bind/DatatypeConverter`

**Comment by marcospereira** [comment]
@Spenhouet just to let you know that we now have releases of sbt-rjs and sbt-jshint for sbt 1.0:

- https://bintray.com/sbt-web/sbt-plugin-releases/sbt-rjs/1.0.10
- https://bintray.com/sbt-web/sbt-plugin-releases/sbt-jshint/1.0.6

**Comment by mkurz** [comment]
`sbt-less` also needs a new release urgently, see https://github.com/sbt/sbt-less/pull/100 - thanks!

**Comment by mkurz** [comment]
I was wrong, there is a recent `sbt-less` release but it looks like the git tag wasn't pushed to remote: https://github.com/sbt/sbt-less/commit/f008176629e89ebeb3882c0d0bba5ded78397f3a#commitcomment-25075088 Sorry for the noise.

**Comment by Spenhouet** [comment]
@marcospereira I tested the compilation with the new sbt-rjs and sbt-jshint release and now I can build with sbt 1.0.2 and Java 9.0.1.

Compiling with sbt 1.0.2 gives me the following warning:

> WARNING: An illegal reflective access operation has occurred
> WARNING: Illegal reflective access by sbt.internal.librarymanagement.ivyint.ErrorMessageAuthenticator$ (file:/C:/Users/spe/AppData/Local/Temp/sbt-global-plugin1stub/boot/scala-2.12.3/org.scala-sbt/sbt/1.0.2/librarymanagement-ivy_2.12-1.0.2.jar) to field java.net.Authenticator.theAuthenticator
> WARNING: Please consider reporting this to the maintainers of sbt.internal.librarymanagement.ivyint.ErrorMessageAuthenticator$
> WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
> WARNING: All illegal access operations will be denied in a future release

Which is resolved but not released: https://github.com/sbt/librarymanagement/issues/169

I can't compile with sbt 1.0.3 (getting multiple errors). Should this work normally?

**Comment by mkurz** [comment]
@Spenhouet sbt 1.0.3 is broken: https://github.com/sbt/sbt/issues/3684 / https://github.com/sbt/sbt/issues/3690 - just use sbt 1.0.2 for now until 1.0.4 is released.

**Comment by Spenhouet** [comment]
I upgraded to sbt 1.0.3, scala 2.12.4 and play 2.6.7 but now I can't build with Java 8 nor with Java 9.

I can start the play application build with java 9 in my IDE but if I push it with `docker:publish` to a docker repository and then run it with `docker run --rm -p 9000:9000 eu.gcr.io/my-docker-image:1.0` I get an `UnsupportedClassVersionError`.

At first I didn't thought this has anything to do with play or any library and therefore opened a stackoverflow question:

https://stackoverflow.com/questions/47018976/running-java-9-application-with-docker-doesnt-work-unsupportedclassversionerr

Since I need to build and run my project with docker I changed back to Java 8 but now I can't build since sbt-sassify 1.4.10 was build with Java 9 and therefore give the same error when building the project with Java 8.

**Comment by richdougherty** [comment]
For those Googling, you can solve the `java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter` problem shown by @markehammons by using `SBT_OPTS="--add-modules java.xml.bind" sbt`.

**Comment by richdougherty** [comment]
I'm taking a look at this issue now.

**Comment by richdougherty** [comment]
Note to self: when this is working update https://wiki.openjdk.java.net/display/quality/Quality+Outreach to include Play.

**Comment by mkurz** [comment]
If you don't want to write `SBT_OPTS=...` in front of every `sbt` call and also don't want to always `export _JAVA_OPTIONS="--add-modules=..."` here is how I solved it:

There are three "levels" were you need to define the modules.

1. For compiling. For that you can put in your `build.sbt`:
```
javacOptions in (Compile, compile) ++= Seq("--release", "9") ++ Seq("--add-modules=module1,module2")
```
In case you use sub-modules you have to defined that javac options for all your sub-modules.

2. When using `sbt run`. To let the `java` process that runs your app in **dev** mode know your modules you can put them into `.jvmopts` inside your project folder. The sbt java process will be launched with that options:
```
# Inside your .jvmopts
--add-modules=module1,module2
```

3. For staging (`sbt stage`), etc. To pass the modules to the generated stage-script you can put this in your `build.sbt` (This should also work when you generate lets say a debian package or docker image etc. - but I didn't try):
```
javaOptions in Universal ++= Seq("-J--add-modules=module1,module2")
```

**Comment by mkurz** [comment]
BTW: I first tried to solely set `javaOptions ++= Seq("-J--add-modules=module1,module2")` but that didn't work for `sbt run` and of course also not for compiling. I also tried to set `fork := true` but that didn't help in dev mode for `run` as well. However it did work for `sbt runProd` and `sbt stage` (see comment above).

**Comment by mkurz** [comment]
It would be great if we could just define our modules at one single place and they will be picked up for compiling, devmode run (`sbt run`) and for distribution (`sbt stage`, docker images, etc.)

**Comment by mkurz** [comment]
One more comment: `sbt runProd` does pick up the `javaOptions` because it just runs `stage` and then spins a new process and [passes that options to that process](https://github.com/playframework/playframework/blob/2.6.7/framework/src/sbt-plugin/src/main/scala/play/sbt/run/PlayRun.scala#L240-L254). That doesn't work for `sbt run` even though when `fork := true` because we don't even extract any `-J` params, but only the `-D` properties from the `javaOptions`, see [here](https://github.com/playframework/playframework/blob/2.6.7/framework/src/run-support/src/main/scala/play/runsupport/Reloader.scala#L157).

**Comment by mkurz** [comment]
BTW: `sbt -jvm-debug <port>` doesn't work yet as well: https://github.com/sbt/sbt-launcher-package/issues/185

**Comment by richdougherty** [comment]
Thanks for the analysis @mkurz. We may have to wait for sbt needs to get a "Java 9" upgrade before this is a smooth experience in Play. For the moment we can fix breaking bugs and provide documentation.

For me, at least, I can get a simple Play app running with the `--add-modules` option mentioned [here](https://github.com/playframework/playframework/issues/7879#issuecomment-344433851). All your notes about the various combinations of `run` are very useful.

**Comment by mkurz** [comment]
The `sbt -jvm-debug <port>` issues has been fixed: https://github.com/sbt/sbt-launcher-package/pull/197 / https://github.com/sbt/sbt-launcher-package/issues/185

**Comment by richdougherty** [comment]
@markehammons, @mkurz, et al: I'd like to organise this issue a bit more so that we can address Java 9 compatibility in a systematic way. Here's my plan, let me know if any objections:

1. Close this ticket because I believe the original issue was fixed by an upgrade to Play 2.6.7 and sbt 1.0.
2. Make a new issue to track Java 9 compatibility at a high level. This issue will stay open until we get Java 9 support working and will contain links to all the sub-issues that we're working on. This issue shouldn't go into detail about issues it should just have links to other issues.
3. Create an issue to track "missing modules", e.g. the `java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter` issue - we have a workaround with command line options but this needs to be at least documented somewhere. @mkurz's [comment above](https://github.com/playframework/playframework/issues/7879#issuecomment-344438326) is a good start.
4. Create an issue to track all the "illegal reflective access" warnings. We can fix these by: (a) adding command line options (b) making Play libraries into modules (c) updating upstream libraries. At the moment these warnings aren't blocking users from using Java 9; they're just annoying.
5. Create an issue to track the work of making Play libraries into modules. As a first step this will involve updating the interplay, play-json, play-ws, Play, Twirl, etc projects to use sbt 1 for their builds so that we can compile with Java 9.
6. Create an issue to track the docker issue raised by @Spenhouet. This may need an upstream change in sbt-native-packager.

**Comment by mkurz** [comment]
Sounds fine to me.

> Create an issue to track the docker issue raised by @Spenhouet. This may need an upstream change in sbt-native-packager.

Which issue? Can you reference?

**Comment by marcospereira** [comment]
@richdougherty agree with all of your points, just want to make some comments.

> Create an issue to track "missing modules", e.g. the java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter issue - we have a workaround with command line options but this needs to be at least documented somewhere. @mkurz's comment above is a good start.

Maybe we can do more than document here. Play can detect Java 9 and add the required modules, at least the ones we need out of the box.

> Create an issue to track the work of making Play libraries into modules. As a first step this will involve updating the interplay, play-json, play-ws, Play, Twirl, etc projects to use sbt 1 for their builds so that we can compile with Java 9.

Can't this be all part of the Java 9 high-level issue?

> Create an issue to track the docker problem raised by @Spenhouet. This may need an upstream change in sbt-native-packager.

If [this answer](https://stackoverflow.com/a/47356730/4600) is correct then it was just a configuration issue.

**Comment by Spenhouet** [comment]
> If this answer is correct then it was just a configuration issue.

Yes, you are right. It was just a configuration issue. Still can't use Java 9 since other libraries are not up-to-date but this has nothing to do with Play.
The issue of this thread and my issues with Play & Java 9 are resolved.
Thank you 👍

**Comment by mkurz** [comment]
Guice also doesn't support Java 9 yet, I just ran into an exception. Here is the Guice Java 9 issue: https://github.com/google/guice/issues/1085

**Comment by richdougherty** [comment]
Link to a Stack Overflow issue that may require a Java 9 upgrade to work: https://stackoverflow.com/a/47858792/49630.

**Comment by mkurz** [comment]
Guice with Java 9 support released: #8259

**Comment by schmitch** [comment]
Actually I guess the xml.bind problem was a sbt problem since I could actually run the master branch on sbt 1.1.2 without problems, the only issue I see is: #8350 (play didn't used anything from xml.bind)

**Comment by eximius313** [comment]
`xml.bind` is not a SBT problem, but JDK Module System.
TL;DR - just add theese to your dependencies:
```
  //For Java9
  "javax.xml.bind" % "jaxb-api" % "2.3.0",
  "javax.annotation" % "javax.annotation-api" % "1.3.2",
  "javax.el" % "javax.el-api" % "3.0.0",
  "org.glassfish" % "javax.el" % "3.0.0",

```

**Comment by dwijnand** [comment]
Should we close this ticket off now?  It's become a bit of a grab-bag of not-strictly-related Java 9+ issues, a bunch of which have been resolved.

@marcospereira @mkurz wdyt?

**Comment by mkurz** [comment]
> Should we close this ticket off now?

@dwijnand OK for me.

**Comment by csaltos** [comment]
you can add JVM options like `--add-opens java.base/java.nio=ALL-UNNAMED` or `--add-opens java.management/sun.management=ALL-UNNAMED` to open the blocked access.

Here's a command to find which module provides which package ->

`java --list-modules | tr @ " " | awk '{ print $1 }' | xargs -n1 java -d`

the name of the module will be shown with the @ while the name of the packages without it

NOTE: tested with JDK 11

IMPORTANT: obviously is better than the provider of the package does not do the illegal access in the first place, but until the transition is done, it is an alternative.
